### ISSUE TITLE
Sync issues with revisions out of order and corrupted

### ISSUE THREAD

**Issue by rezzap** [Automattic/Simplenote-United#80]
We have a user advising on the following two issues: 

1- Notes will randomly convert back to an old version, and when changed so its the latest version will suddenly go back to that old version

2- Text next to each other will get chewed up or duplicated randomly on top of each other

When asked to clarify some details they shared the following:

> Normally when using Simplenote app, I make some changes, and then close the app. Same on browser, I'll edit it and close the browser when I'm done. The correct version does in fact show up in history, kind of. At time it'll have the chewed up data I mentioned in (2.)
> 
> Haven't seen the problem attached to a specific platform. Except I guess the desktop app, but that may of been cus it was the first time I was noticing the problem.
> 
> While I can't really say how to reproduce it, I'll give a pipeline on how I use the app on a regular day.
> 
> Read/write on smartphone device & using browser on laptop throughout the day -> Read/write on Desktop and smartphone device by end of day.
> 
> Then either the next day or in a few days I'll check back on those notes on both my smartphone and browser and see that the note was rerouted to an old date. Then I have to go into history and change it back to the last time I changed it hopefully with minimal data getting chewed.
> 

I thought these might be related to #992 or #1787, but the additional details make me think that's not the case. 

Device details:
1. Desktop Computer, Smart phone device
2. Windows 10 and Android
3. Version 2.4
4. Google Chrome

Will ask for further details, but the user is also going to try to see if they can catch the exact steps to replicate next time it happens.

Reported in 2797007-zen

**Comment by dmsnell** [comment]
> I make some changes, and then close the app.

Can we get a feel for how quickly after making edits they close the app? There's a timeout of a couple seconds before we synchronize edits with the server. This is there to prevent a proliferation of network activity and to prevent flooding the revision history.

If someone closes the app right after making changes then it might hold on to those edits until the next time it's opened. At that point other devices may have made updates to the same document but since the previous edits never synchronized it will apply them on top of anything that changed in the meantime.

In short: if not doing this already, it's best to wait a few seconds before closing the app/browser after making edits.

**Comment by rezzap** [comment]
We have another report of this using the MacOS app and the iOS app.

Mac OS 10.14.6 on Macbook pro
- Simplenote 1.9.1

iOS 13.3 on iPhone 6
- Simplenote 4.16

> On desktop, rather than showing the updates I made on my phone at 10:33AM it shows me the previous state from desktop at 10:10AM. but I can revert to 10:33AM if I go to history, which seems strange.

I'm asking for more detailed steps to reproduce and also on what happens when they do try to restore that revision. My understanding is that the changes they make on MacOS don't sync with the iOS app, and changes made there get lost if they restore. 

2814151-zen

**Comment by sarahfg** [comment]
Here's the user's details on how they are using the app:

```
New text sometimes appear(again=restored) and sometimes disappear, but in most cases keep disappearing when using PDA( android7.1.1).
After the data disappear, it is not remained even in the edit history nor shown on other devices(web app/ iOS).

To avoid this data disappearing problem, I manage by 5 steps below

1.copy text to clip board (ctrl+a, ctrl+c)
2.exit simlenote by app switcher button and re-enter simplenote.
3.Sync starts
4.if text disappears after Sync, paste text (ctrl+v) and repeat step.2
5.Sync starts again

Therefore, after inputting every single line I have to copy the text by hand(ctrl+a, ctrl+c), several times in a minute.
Consequently, I sometimes delete all sentence in a note by myself due to keyboard miss-touch (ctrl+a, ctrl+z!)
```

#2830573-zen

**Comment by sarahfg** [comment]
Using the web app solves their issue. It still persists on 2.5 in Android App however

**Comment by rezzap** [comment]
I have another report of this in 2878434-zen

using Simplenote on iOS 13.4.1, they think it's mostly related to the iPad. 

It seems to be happening when they go back and forth between the app and other programs on the same device. 

They mention the following prevents it from happening:

> There may be a way that I’ve found to keep it from happening. Rather than tapping <Notes to go back after I’ve entered or edited a note, I first hide the keyboard. So maybe the app is doing something smart when it knows I’ve stopped editing that way, but the app isn’t so smart when I stop editing by going back to the list of notes?

It sounds a bit similar to the issue on Android where the notes are lost when using the navigation drawer buttons in https://github.com/Automattic/simplenote-android/issues/992

**Comment by codebykat** [comment]
User in 2878434-zen updated:

> It turns out I had left Simplenote open in a Safari window on my Mac (running Mojave). [...] I also haven't seen the problem now that I’ve closed that window. In the past, I’ve always immediately closed the browser window after using Simplenote on my laptop—I use it much more often in iOS, where I of course close the app when finished.

**Comment by sshaikh** [comment]
I've also had a few out of order edits between Android and Electron:

1) add 10 lines via electron, verify up to date note on Android (v1)
2) add lines on android. press power button on phone, switch out (etc), v2
3) delete different lines on electron, v3
4) open android app. still see v2. this then gets pushed as the latest version, v4.
5) see v4 (ie v2) on electron. history shows all versions including the clobbered v3.
6) restoring v3 puts everything back.

I've had various issues like these with Simplenote up to the point where I am now anxious to touch any notes on a different device until I've done a visual inspection that the note has transferred okay.

As many others have said, this seems to be easily solvable with a "sync now" button or even a "last saved" timestamp, as at least then you can stop what you're doing until a resolution is found. These dates seem to be available in the history browser where they do come in useful when fixing these kind of mixups.

Unfortunately silent automatic magical syncs just doesn't work on the edges, and is painful when it goes wrong.

**Comment by erricgunawan** [comment]
Probably related from 3111862-zen:

> After I completed taking notes from the browser version of Simplenote, I come back today and that particular note is missing 95% of the notes that I took.

> It appears it only recorded just the beginning of the notes that I took.

> I looked in my mobile app (Android) and the same note in question was a full page of a repeated line of the original note.

> This has happened multiple times now.

I'm collecting the user's device and Simplenote app info now.

**Comment by codebykat** [comment]
We hope and believe most of these sync issues have been fixed in the latest version of web/Electron. We've also added a last synced time there. If folks are still having trouble let's try to narrow it down to specific platforms and open issues in the relevant repository (I know some work is also being done over in simplenote-android on reports of syncing problems).
